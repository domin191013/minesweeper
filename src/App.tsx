import React from "react";
import "./App.css";
import GameMain from "./Game";

function App() {
  return <GameMain />;
}

export default App;
