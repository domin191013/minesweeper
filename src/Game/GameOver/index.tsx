import { Alert } from "@mui/material";

interface IProp {
  gameOver: number;
}

function GameOver(props: IProp) {
  const { gameOver } = props;

  return (
    <div>
      {gameOver === 1 ? (
        <Alert severity="success">Congratulations! You won!</Alert>
      ) : gameOver === 2 ? (
        <Alert severity="error">You failed! Please try again!</Alert>
      ) : (
        <span></span>
      )}
    </div>
  );
}

export default GameOver;
