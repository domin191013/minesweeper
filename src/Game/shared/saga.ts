import { EventChannel, eventChannel } from "redux-saga";
import { call, put, take, fork, ChannelTakeEffect } from "redux-saga/effects";
import {
  CONNECT_SAGA_PUT,
  START_LEVEL,
  GAME_NOT_STARTED,
  PAINT_MAP,
  OPEN_CELL,
  GAME_ENDED
} from "./constants";

const socketServerURL = "wss://hometask.eg1236.com/game1/";

function connect() {
  const socket = new WebSocket(socketServerURL);

  return new Promise((resolve, reject) => {
    socket.onopen = () => resolve({ socket });
    socket.onclose = () => reject(new Error("ws:connect_failed "));
  }).catch((error) => ({ socket, error }));
}

function createSocketChannel(socket: WebSocket): EventChannel<any> {
  return eventChannel((result) => {
    const reconnectHandler = (ping: number) => {
      result({
        type: CONNECT_SAGA_PUT,
        payload: ping,
      });
    };

    const reconnectingHandler = (ping: number) => {
      result({
        type: CONNECT_SAGA_PUT,
        payload: ping,
      });
    };

    socket.onmessage = (message: any) => {
      const data = message.data;
      const command = (data || "").split(":")[0];
      switch (command) {
        case "map": {
          if (data.indexOf("Not started") >= 0) {
            result({
              type: GAME_NOT_STARTED,
              payload: null,
            });
            console.log("Game is not started!");
            break;
          }
          const map = data.trim().substr(5).split("\n");
          result({
            type: PAINT_MAP,
            payload: map,
          });
          console.log("Painting a new map!");
          break;
        }
        case "open": {
          if (data.indexOf("You lose") >= 0) {
            console.log("%c Game Over!", "color: red;");
            result({
              type: GAME_ENDED,
              payload: "lose",
            });
          } else if (data.indexOf("You win") >= 0) {
            console.log("%c You won!", "color: green;", data);
            result({
              type: GAME_ENDED,
              payload: "won",
            });
          }
          break;
        }
        case "help":
        case "new": {
          break;
        }
        default: {
          console.log(
            "%c warning: command not handled yet",
            "color: orange;",
            command,
            data
          );
          break;
        }
      }
    };
    const unsubscribe = () => {};
    return unsubscribe;
  });
}

// watch if the game starts
function* watchStartGame(socket: WebSocket) {
  while (true) {
    const { payload } = yield take(START_LEVEL);
    console.log("Saga - new start game request");
    socket.send(`new ${payload}`);
    socket.send("map");
  }
}

// watch if the cell is opend
function* watchOpenCell(socket: WebSocket) {
  while (true) {
    const { payload } = yield take(OPEN_CELL);
    console.log("Saga - open cell request");
    socket.send(`open ${payload}`);
    socket.send("map");
  }
}

function* watchTask(socket: WebSocket) {
  yield fork(watchStartGame, socket);
  yield fork(watchOpenCell, socket);
}

function* putAction(socket: WebSocket) {
  const event: EventChannel<any> = yield call(createSocketChannel, socket);
  while (true) {
    let action: ChannelTakeEffect<any> = yield take(event);
    console.log("flow action: ", action);
    yield put(action);
  }
}

function* task(socket: WebSocket) {
  yield fork(putAction, socket);
  yield fork(watchTask, socket);
}

export function* socketRootSaga(): Generator<any, any, any> {
  const { socket, error } = yield call(connect);

  if (!error) {
    console.log("It's now connected to server");
    socket.send("help");
    yield put({ type: CONNECT_SAGA_PUT, payload: "Connected" });
    return yield fork(task, socket);
  } else {
    console.log("The connection is disconnected");
    put({ type: CONNECT_SAGA_PUT, payload: "Disconnected" });
    return null;
  }
}
