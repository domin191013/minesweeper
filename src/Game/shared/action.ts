import { IActionTypes } from "./action.types";
import { START_LEVEL, OPEN_CELL } from "./constants";

export function startLevel(level: string): IActionTypes {
  console.log("START LEVEL", " - ", level);
  return {
    type: START_LEVEL,
    payload: level,
  };
}

export function openCell(cell: string): IActionTypes {
  console.log("OPEN LEVEL", " - ", cell);
  return {
    type: OPEN_CELL,
    payload: cell,
  };
}