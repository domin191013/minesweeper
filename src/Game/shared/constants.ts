import { startLevel, openCell } from "./action";
export const CONNECT_GAME = "CONNECT_GAME";
export const START_LEVEL = "START_LEVEL";
export const GAME_NOT_STARTED = "GAME_NOT_STARTED";
export const PAINT_MAP = "PAINT_MAP";
export const OPEN_CELL = "OPEN_CELL";
export const GAME_ENDED = "GAME_ENDED";

export const CONNECT_SAGA_PUT = "CONNECT_SAGA_PUT";

export interface IState {
  map: Array<any>;
  gameStarted: boolean;
  currentLevel: string;
  gameOver: number; //0: not yet ended, 1: won, 2: lose
  message: string;
}

interface IDispatch {
  startLevel: typeof startLevel;
  openCell: typeof openCell;
}

export type IProp = IState & IDispatch;