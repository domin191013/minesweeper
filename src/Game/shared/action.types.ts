import { Action } from "redux";
import { START_LEVEL, OPEN_CELL } from "./constants";

interface IStartLevel extends Action {
  type: typeof START_LEVEL;
  payload: string;
}

interface IOpenCell extends Action {
  type: typeof OPEN_CELL;
  payload: string;
}

interface IActionType extends Action {
  type: string;
  payload: any;
}

export type IActionTypes = IActionType;
