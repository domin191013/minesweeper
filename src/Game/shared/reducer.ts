import { IState } from "./constants";
import { IActionTypes } from "./action.types";
import {
  START_LEVEL,
  GAME_NOT_STARTED,
  PAINT_MAP,
  GAME_ENDED,
} from "./constants";

const initState: IState = {
  map: [],
  gameStarted: true,
  currentLevel: "0",
  gameOver: 0,
  message: "",
};

export default function Reducer(
  state: IState = initState,
  action: IActionTypes
): IState {
  const newState = Object.assign({}, state);
  switch (action.type) {
    case START_LEVEL:
      newState.gameOver = 0;
      newState.currentLevel = action.payload;
      break;

    case GAME_NOT_STARTED:
      newState.gameStarted = false;
      newState.map = [];
      break;

    case PAINT_MAP:
      newState.map = action.payload;
      newState.gameStarted = true;
      break;

    case GAME_ENDED:
      if (action.payload === "won")
        newState.gameOver = 1;
      else
        newState.gameOver = 2;
      break;
  }
  return newState;
}
