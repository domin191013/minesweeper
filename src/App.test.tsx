import { Provider } from "react-redux";
// import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { startLevel, openCell } from "./Game/shared/action";
import App from "./App";
import GameOver from "./Game/GameOver";

const mockStore = configureStore([]);

describe("Minesweeper Application Component", () => {
  let store: any;
  let component: any;

  beforeEach(() => {
    store = mockStore({
      map: [],
      gameStarted: true,
      currentLevel: "0",
      gameOver: 0,
      message: "",
    });
    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  // Testing Menu component
  it("should dispatch startlevel action when click on beginner menu item", () => {
    renderer.act(() => {
      component.root.findAllByType("button")[0].props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(startLevel("1"));
  });

  it("should dispatch startlevel action when click on intermediate menu item", () => {
    renderer.act(() => {
      component.root.findAllByType("button")[1].props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(startLevel("2"));
  });

  it("should dispatch startlevel action when click on advanced menu item", () => {
    renderer.act(() => {
      component.root.findAllByType("button")[2].props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(startLevel("3"));
  });

  it("should dispatch startlevel action when click on expert menu item", () => {
    renderer.act(() => {
      component.root.findAllByType("button")[3].props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(startLevel("4"));
  });
});
// test("renders learn react link", () => {
//   render(<App />);
//   const linkElement = screen.getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });
