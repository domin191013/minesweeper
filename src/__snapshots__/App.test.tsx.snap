// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minesweeper Application Component should render with given state from Redux store 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1ewlecp-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
  >
    Welcome to minesweeper game!
  </div>
  <div
    className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
  >
    <h2>
      Select the level and enjoy the game!
    </h2>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-12fe4sh-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-5t3b3g-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Beginner
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Intermediate
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Advanced
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-ke5b6m-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Expert
        </button>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1dqzm9l-MuiGrid-root"
  >
    <div>
      <span />
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1dqzm9l-MuiGrid-root"
  >
    <table
      style={
        Object {
          "display": "block",
        }
      }
    >
      <tbody />
    </table>
  </div>
</div>
`;
